leaderelection/src/main/java/io/scalecube/leaderelection/MessageListener.java

```diff
@@ -0,0 +1,100 @@
+package io.scalecube.leaderelection;
+
+import io.scalecube.cluster.ICluster;
+import io.scalecube.transport.Message;
+
+import rx.functions.Action1;
+import rx.functions.Func1;
+
+/**
+ * Created by ronenn on 9/13/2016.
+ */
+public abstract class MessageListener {
+
+  enum ListenType {
+    GOSSIP, TRANSPORT, GOSSIP_OR_TRANSPORT
+  }
+  
+  public final ICluster cluster;
+
+  public MessageListener(ICluster cluster) {

```

Comment:
A lot of code duplication in this class.
As I understand  here should be two filter function - "qualifier equal" and "qualifier start with" which pure function and can be defined and saved in the variable and two cases cluster.listen() and cluster.listenGossips() which both return Observable which should be filtered with correct function based on ListenType. Please correct me if I miss something?